sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEUtils.scala

```diff
@@ -57,4 +58,24 @@ object AQEUtils {
       }
     case _ => Some(UnspecifiedDistribution)
   }
+
+  // Analyze the given plan and calculate the required ordering of this plan w.r.t. the
+  // user-specified sort.
+  def getRequiredOrdering(p: SparkPlan): Seq[SortOrder] = p match {
+    case f: FilterExec => getRequiredOrdering(f.child)
+    case c: CollectMetricsExec => getRequiredOrdering(c.child)
+    // We do not need to care whether the sort is global or not, since the output partitioning
+    // is ensured by requiredDistribution.
+    case s: SortExec => s.outputOrdering
+    case p: ProjectExec =>
+      val requiredOrdering = getRequiredOrdering(p.child)
+      // avoid case `df.sort(a, b).select(c)`
+      if (requiredOrdering.forall(e => p.projectList.exists(_.semanticEquals(e)))) {
+        requiredOrdering
```

Comment:
[question] can we handle aliasing as well. 
 For ex :  `df.sort(a, b).select(col('a').as('x'))`

Reply:
This case has already converted which do not need to add back sort. Added some tests.